### DISCUSSION TITLE
Missing write readiness notifications using Child with Stdio::piped?

### DISCUSSION THREAD

**Discussion by adamreichold** [tokio-rs/tokio#5644] [Q&A] [answered]
I have a program that among other things spawns subprocess (which are usually Python scripts) and communicates with them over pipes using a line-oriented protocol. This protocol is half-duplex, i.e. at each point in time only one of the two processes is expected to write a line and the other is expected to read one.

This seems to mostly work, but sometimes the program and its subprocess deadlock. Adding timeouts on the Rust/tokio/parent side, I found out that it was the writing which blocked indefinitely, e.g. `stdin.write_all(&buf).await?`. Killing a blocked Python script using `SIGINT` and looking at the resulting backtrace, I determined that it was blocking in `sys.stdin.readline()`.

This does not really make sense to me as when the pipe buffer is full, the read should succeed and empty it. However, increasing the kernel's pipe buffer size to 1 MB which is larger than all lines I am trying to write so that all writes should succeed immediately did apparently resolve the issue. To me, this appears like the write readiness notifications when the pipe buffer is significantly smaller than the lines I try to write are just lost. (While the timeouts are currently set to 60s, the deadlocks really are deadlocks persisting for days if not resolved manually or via the timeouts.) 

Of course, a bug in my code is much more likely, but at this point I have run out of ideas trying to diagnose it. Also due to the protocol being half-duplex, the classic case of pipe buffer deadlock involving two pipes should not happen and indeed continuously draining `stdout` in a separate task while writing to `stdin` did not affect the deadlocks. I would be glad if anyone could offer additional insight on why my writes are blocking indefinitely.

To add some context, the parent code is [here](https://gitlab.opencode.de/umwelt-info/metadaten/-/blob/main/src/harvester/external.rs) and the child code is [here](https://gitlab.opencode.de/umwelt-info/metadaten/-/blob/main/deployment/external_harvesters/website-bfn.py) and [here](https://gitlab.opencode.de/umwelt-info/metadaten/-/blob/main/deployment/external_harvesters/common.py). I am sorry for not having a more minimal reproducer, but I have a hard time reproducing it at all. For example, it seems to happen only on our staging VM running Ubuntu 22.04 and hence kernel 5.15, but not on our development VM currently running Ubuntu 23.04 and hence kernel 6.2. Due to the kernel version differences, I wonder if [the "recent" rework](https://lwn.net/Articles/864947/) of the kernel's pipe subsystem have anything to with it, but that is admittedly a long shot.

EDIT: To make this more concrete, when they deadlock, the parents waits [here](https://gitlab.opencode.de/umwelt-info/metadaten/-/blob/49e2ac9a30f4d501c3d99a894f6e32418210c50b/src/harvester/external.rs#L140) and the child blocks [here](https://gitlab.opencode.de/umwelt-info/metadaten/-/blob/49e2ac9a30f4d501c3d99a894f6e32418210c50b/deployment/external_harvesters/common.py#L20).

EDIT: Trying a minimal reproducer which just ping-pongs 500 kB worth of line data between a Rust parent and Python child in a tight loop does not seem to trigger the issue. So at least some of the other work for the runtime seems required.

**Comment by adamreichold** [comment]
I am testing this against the 5.19 HWE kernel shipped with Ubuntu 22.04 now. Maybe the problem magically disappears that way...

**Reply by adamreichold** [reply]
But it did not so it is not some interaction with how the kernel notifies on I/O readiness of pipes. This was admittedly pretty unlikely from the start.

**Comment by adamreichold** [comment]
I fear this could be an issue in tokio or the underlying I/O stack after all: I replaced that single call to `write_all` by its blocking equivalent using `block_in_place` as in [this commit](https://gitlab.opencode.de/umwelt-info/metadaten/-/commit/6de221fd258a1ea6d63146396d3496982fa80ce8) and the problem has not re-appeared for a week even with the default kernel pipe buffer size of 64 kB where it would appear regularly every two to three days.

I would be grateful for any insights or advice on what I should collect to turn this into an actionable bug report.

**Reply by Darksonn** [reply]
It sounds very weird. I've looked at your report a few times, but I have no idea what could be wrong. I'm pretty sure we use edge triggering in the "correct" way for pipes too, and we should be able to tolerate extra wakeups, so the kernel rework doesn't sound like the issue, but if kernel versions affect it, then I'm not sure what else it could be.

If you're going to compare with a synchronous version, I would probably keep the buffering mechanism the same to keep the differences minimal.

**Reply by adamreichold** [reply]
> If you're going to compare with a synchronous version, I would probably keep the buffering mechanism the same to keep the differences minimal.

That's true, I went for `BufWriter` since I did not need the intermediate buffer for the synchronous version but indeed this would change the system calls issued. I will deploy writing the full buffer at once to check if this still seems to resolve the issue...

**Reply by adamreichold** [reply]
I will deploy a [more minimal change](https://gitlab.opencode.de/umwelt-info/metadaten/-/commit/af8ccc8fbc4050a1bbe3ceea921c80544baa5291) and see whether this resolves the issue. Of course, this will take a few days of observation to be reasonably sure as it was always an intermittent problem...

**Reply by adamreichold** [reply]
The above-linked minimal change has been running for a week _without_ the issue reoccurring. I think this strongly suggests that the implementation of `AsyncWrite` or `AsyncWriteExt` has an issue when using pipes at least on Ubuntu 22.04 LTS or its HWE kernel.

As workaround for us, I will switch to `std::process::Comand` for now as it means I can avoid the unsafe code required for the minimal workaround and also allows me to reduce the buffering as I can use the integration of `serde_json` with `io::{Read, Write}`.

Long-term, I would be glad though to find the root cause for this to ideally provide a fix for Tokio or its underlying I/O stack or at least identify the system configuration which prevents this from working. I am willing to do more experimentation to figure this out and would be glad for any pointers on how to proceed usefully.

**Reply by Darksonn** [reply]
Unfortunately, I don't really know.

**Reply by adamreichold** [reply, marked as answer]
Could this be an instance of #6133 ?

**Reply by adamreichold** [reply]
I have deployed a change which does nothing but bump Tokio to 1.34 and switch back to the asynchronous version of `Command` this is now running without issues for week. If the problem does not resurface for another one, I would declare this as resolved as another instance of #6133.

**Comment by adamreichold** [comment]
Seemed to have been an instance of #6133 after all.

**Comment by beroal** [comment]
> However, increasing the kernel's pipe buffer size to 1 MB which is larger than all lines I am trying to write so that all writes should succeed immediately did apparently resolve the issue.

May [line buffering](https://man7.org/linux/man-pages/man3/setbuf.3.html#DESCRIPTION) be the reason?

**Reply by adamreichold** [reply]
As written in

> Seemed to have been an instance of https://github.com/tokio-rs/tokio/issues/6133 after all.

this was an effect of #6133 and bumping Tokio to a version with the fix resolved the issue.
